### PR TITLE
Add ignore case option

### LINKED ISSUE #11: Adding an ignoreCase option
I started work on adding an ignoreCase option.
First of all, that was a fun trail down the chain to find out how `fkill` and its dependencies work. I quickly stopped searching the npm registry for them and searching Sindre's personal repo's instead 😄 .
Secondly: the fkill logo is _beautiful_. Had to be said.
Third: adding an ignoreCase option turns out to be a real pain.

**Linux**
Easy. [Add '-I' flag](https://linux.die.net/man/1/killall). Done. Go Linux.

**macOS and BSD**
No ignore case option. The closest provided option is using a regular expression 😓 .
Both do however have `pgrep` and `pkill`, both of which, feature an ignore case option!

**SunOS**
What even is this 😂 ? Anyway [not even a little bit of options](http://www.manpages.info/sunos/killall.1.html)

**Windows**
[Zero options](https://ss64.com/nt/taskkill.html) to ignore case, zero options to [collect pid's](https://ss64.com/nt/tasklist.html) with ignore case.

We could just issue two kill commands.. Far easier. Less correct.

Please let me loose on this :3 . If you say, _far_ too much hassle for ignore case, don't want to complicate the code that bad, or take on an abstraction for it .. I totally understand.

### PR DESCRIPTION
**Pull request by AlexTes**
Adds an ignore case option using suggestions from #11.

* [x] Add ignore case test
* [x] Add bad option plus input combination test (currently best effort solution)
* [x] Add macOS / BSD support
* [x] Add linux support
* [x] Add windows support
* [x] Update documentation

Question: how do you feel about these checklist PRs? I can imagine they're very inconvenient notification-wise.

---
Fixes #11

### CODE REVIEW

**Comment by AlexTes** [comment]
I guess Travis is Linux. How will we test the other OS's?

**Comment by SamVerschueren** [comment]
Travis supports macOS as well if I'm not mistaken. For Windows we can use AppVeyor.

File: index.js [hunk 1]
```diff
@@ -12,6 +12,21 @@ function win(input, opts) {
 	});
 }
 
+function macOSKill(input, opts) {
+	const cmd = typeof input === 'string' ? 'pkill' : 'kill';
+	const args = [input];
+
+	if (opts.force) {
```
Comment (AlexTes) [inline index.js]: All common prepping of args could be abstracted 🤔. Keep in mind there are three combinations of environments that might share a common `args` mutation based on the same option, that the third environment does not. I feel for now the complexity of an abstraction outweighs the code duplication, but I trust your judgment more on this.

File: index.js [hunk 2]
```diff
@@ -25,10 +40,17 @@ function def(input, opts) {
 
 module.exports = (input, opts) => {
 	opts = opts || {};
-
-	const fn = process.platform === 'win32' ? win : def;
 	const errors = [];
 
+	let fn;
```
Comment (AlexTes) [inline index.js]: Are you guys fine with this or do you prefer something like `let fn = () => Promise.reject('Unsupported environment');`?

File: index.js [hunk 3]
```diff
@@ -12,23 +12,51 @@ function win(input, opts) {
...
 }
 
-function def(input, opts) {
-	const cmd = typeof input === 'string' ? 'killall' : 'kill';
+function macOSKill(input, opts) {
+	const killByName = typeof input === 'string';
+	const cmd =  killByName ? 'pkill' : 'kill';
 	const args = [input];
 
 	if (opts.force) {
```
Comment (AlexTes) [inline index.js]: All common prepping of args could be abstracted 🤔. Keep in mind there are three combinations of environments that might share a common args mutation based on the same option, that the third environment does not. I feel for now the complexity of an abstraction outweighs the code duplication, but I trust your judgment more on this.

**Comment by AlexTes** [comment]
I would really appreciate an opinion on the 'best effort' solution to the bad input plus option scenario. I personally just don't have the experience to have a strong opinion on this. But what would you say should happen if the user inputs something he clearly didn't mean and is impossible to execute. Like passing a `pid` with the `ignoreCase` option.

**Comment by AlexTes** [comment]
Wrestled quite a bit with the 'winKill'. I've run out of time for tonight but I'll soon add the control flow I'm looking for but failing to neatly build. For now, I've just added the simplest solution I could come up with. It helps discuss options I think.

File: index.js [hunk 4]
```diff
@@ -1,34 +1,74 @@
...
 }
 
-function def(input, opts) {
-	const cmd = typeof input === 'string' ? 'killall' : 'kill';
+function winKill(input, opts) {
+	const killByName = typeof input === 'string';
+	if (killByName && opts.ignoreCase) {
+		return tasklist()
+			.then(tasks => tasks.filter(task => task.imageName.toLowerCase() === input.toLowerCase()))
+			.then(names => Promise.all(names.map(name => winKillProcess(name, opts))));
```
Comment (AlexTes) [inline index.js]: `taskkill` also supports an array of strings as input. I choose not to use it to keep the function more consistent with the rest.

File: test.js [hunk 5]
```diff
@@ -6,14 +6,20 @@ import processExists from 'process-exists';
 import delay from 'delay';
 import m from './';
 
-// Ensure the noop process has time to exit
-const noopProcessExitDelay = 10;
+async function noopProcessExists(t, expectExists, pid) {
```
Comment (AlexTes) [inline test.js]: How do we feel about a test helper like this?

File: package.json [hunk 6]
```diff
@@ -38,7 +38,8 @@
     "aggregate-error": "^0.1.0",
     "arrify": "^1.0.0",
     "execa": "^0.2.1",
-    "taskkill": "^2.0.0"
+    "taskkill": "^2.0.0",
+    "tasklist": "^2.1.0"
```
Comment (sindresorhus) [inline package.json]: `^3.0.0`

File: readme.md [hunk 7]
```diff
@@ -56,6 +56,13 @@ Default: `false`
 
 Force kill the process.
 
+##### ignoreCase
+
+Type: `boolean`<br>
+Default: `false`
+
+Ignore capitalization when killing a process.
```
Comment (sindresorhus) [inline readme.md]: > Ignore capitalization when matching a process.

?

File: test.js (hunk 5 above)
Comment (sindresorhus) [inline test.js]: 👍

File: test.js [hunk 8]
```diff
@@ -6,14 +6,20 @@ import processExists from 'process-exists';
...
+	// Ensure the noop process has time to exit
+	await delay(10);
+	if (expectExists) {
+		return t.true(await processExists(pid));
+	}
+
+	return t.false(await processExists(pid));
```
Comment (sindresorhus) [inline test.js]: I don't see the point in returning this.

Comment (sindresorhus) [inline test.js]: Could just be:

```js
const fn = expectExists ? t.true : t.false;
fn(await processExists(pid));
```

File: test.js (hunk 5 above)
Comment (sindresorhus) [inline test.js]: I would order it: `t`, `pid`, `expectExists` though.

Comment (sindresorhus) [inline test.js]: Seems like `expectExists` is always `false`. ?

File: test.js [hunk 9]
```diff
@@ -46,3 +51,17 @@ if (process.platform === 'win32') {
 		}
 	});
 }
+
+test('ignore case', async t => {
+	const pid = await noopProcess({title: 'Capitalized'});
+	await m('capitalized', {ignoreCase: true});
+
+	noopProcessExists(t, false, pid);
```
Comment (sindresorhus) [inline test.js]: You need `await` here.

File: test.js [hunk 10]
```diff
@@ -46,3 +51,17 @@ if (process.platform === 'win32') {
...
+});
+
+test('ignore ignore-case for pid', async t => {
+	const pid = await noopProcess();
+	await m(pid, {force: true, ignoreCase: true});
+
+	noopProcessExists(t, false, pid);
```
Comment (sindresorhus) [inline test.js]: You need `await` here.

File: test.js (hunk 5 above)
Comment (AlexTes) [inline test.js]: True, I overcomplicated for no good reason. It's out, whoever needs it can add it back 😄.

**Comment by AlexTes** [comment]
Hm, doesn't seem to be the delay 😫. I'll do some local windows testing.

**Comment by MarkTiedemann** [comment]
It seems like `noop-process` with the `title` option doesn't work on Windows.

```js
// noop.js
require('noop-process')({title: 'noop-process'}).then(console.log)
```

```
λ node noop
11888
```

```
λ tasklist /fi "pid eq 11888" /v

Image Name                     PID Session Name        Session#    Mem Usage Status          User Name                                              CPU Time Window Title

========================= ======== ================ =========== ============ =============== ================================================== ============ ========================================================================
node.exe                     11888 Console                    3     19,584 K Unknown         C-3PO\mark                                              0:00:00 N/A
```

Instead of `noop-process`, the window title is `N/A` (not available). Someone should probably open an issue in that repo. :)

Then again, that is irrelevant I guess, because in the test, you were killing based on the image name, not based on the window title.

----------------

I also saw the following code in your `index.js`:

```js
...
return tasklist()
    .then(tasks => tasks.filter(task => task.imageName.toLowerCase() === input.toLowerCase()))
    .then(names => Promise.all(names.map(name => winKillProcess(name, opts))));
...
```

Did you know that `taskkill` (just like `tasklist`) is already case insensitive by default?

```
λ tasklist /fi "imagename eq BOINC.EXE"                                      
                                                                             
Image Name                     PID Session Name        Session#    Mem Usage 
========================= ======== ================ =========== ============ 
boinc.exe                     1524 Console                    3     10,832 K 
```
```                                                                                                                  
λ taskkill /im BOINC.EXE /f                                                  
SUCCESS: The process "boinc.exe" with PID 1524 has been terminated.
```

**Comment by AlexTes** [comment]
First off, thanks @MarkTiedemann! Your pointers are very helpful.

The only reason the other test with a title on windows seems to work is because it spawns the actual windows notepad. Not sure how that even works on AppVeyor. In any case, that makes any other test using the same trick share state which messes up the tests. I'll try some alternatives.

**Comment by sindresorhus** [comment]
@AlexTes Still interested in finishing this up?

**Comment by AlexTes** [comment]
@sindresorhus yes 😓. going to make an attempt tomorrow. do feel free to close the issue if it ends up stale again 🙁.

**Comment by AlexTes** [comment]
New laptop, took a while to get windows set up, made some progress. Don't pull the plug yet :smile:.

**Comment by sindresorhus** [comment]
@AlexTes Should I close? :)

**Comment by AlexTes** [comment]
@sindresorhus you should've a long time ago, you're too kind 😉 .

I just put another two hours in. Windows ignores case so the ignore case feature has no effect on windows. I really wanted to test it to work on windows ~but I just can't find a way~.

The image name seems to be derived from the actual command executed. So unless you have an actual command to execute, like 'notepad.exe'. You're out of luck for a title based test on windows.

**Comment by AlexTes** [comment]
😲 ✨

**Comment by AlexTes** [comment]
coollikethat

**Comment by sindresorhus** [comment]
Yay! Glad to see this implemented. Nice work @AlexTes :)

**Pull request merged**